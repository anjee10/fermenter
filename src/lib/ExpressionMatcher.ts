import { CucumberExpression, ParameterTypeRegistry, RegularExpression } from 'cucumber-expressions';

import { IMatch } from '../types';

/**
 * Produces an expression matcher instance
 * from the cucumber-expressions library
 */
export function ExpressionMatcher (match: IMatch, registry?: ParameterTypeRegistry) {
  if (!registry) {
    registry = new ParameterTypeRegistry();
  }

  const matcher = typeof match === 'string'
    ? new CucumberExpression(match, registry)
    : new RegularExpression(match, registry);

  return matcher;
}
