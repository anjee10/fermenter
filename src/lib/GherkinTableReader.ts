import { IGherkinAstTableRow, IGherkinTableParam } from '../types';

/**
 * A gherkin feature file's rows:
 *
 *   | a | b |
 *   | 1 | 2 |
 *   | 3 | 4 |
 *
 * Becomes `IGherkinAstTableRows[]`
 *
 * @example
 *
 * const table = GherkinTableReader({ rows })
 */
export function GherkinTableReader ({ rows: inputRows = [] }: {
  rows: IGherkinAstTableRow[],
}): IGherkinTableParam {
  const rowValues = inputRows
    .map(({ cells }) => cells.map(({ value }) => value));

  const getLeft = () => rowValues.map((cells) => cells[0]);
  const getTop = () => rowValues[0];

  const dict = <IGherkinTableParam['dict']> (() => {
    const keys = headers();

    const contentRows = rowValues.slice(1);

    const tuples = keys.map((key, index): [string, string[]] => {
      return [
        key,
        contentRows.map((cells) => cells[index]),
      ];
    });

    return new Map(tuples);
  });

  dict.left = () => {
    const keys = rowValues.map((cells) => cells[0]);

    const mapTuples = rowValues.map((cells, index): [string, string[]] => {
      /** The cells, after the left-most */
      const contentCells = cells.slice(1);

      return [keys[index], contentCells];
    });

    return new Map(mapTuples);
  };

  dict.matrix = () => {
    const topKeys = getTop()
      .slice(1); // Skip blank

    const leftKeys = getLeft()
      .slice(1); // Skip blank

    const contentRows = rowValues
      .slice(1) // Skip top
      .map((cells) => cells.slice(1)); // Skip left

    const leftTuples = leftKeys.map((leftKey, leftIndex): [string, Map<string, string>] => {
      const topTuples = topKeys.map((topKey, topIndex): [string, string] => {
        const cells = contentRows[topIndex];

        return [topKey, cells[leftIndex]];
      });

      return [leftKey, new Map(topTuples)];
    });

    return new Map(leftTuples);
  };

  const rows = <IGherkinTableParam['rows']> (() => {
    return inputRows.map(({ cells }) => cells.map(({ value }) => value));
  });

  rows.mapByTop = () => {
    const keys = getTop();

    const contentRows = rowValues.slice(1);

    return contentRows.map((cells) => {
      const tuples = keys.map((key, index): [string, string] => {
        return [key, cells[index]];
      });

      return new Map(tuples);
    });
  };

  const headers = <IGherkinTableParam['headers']> (() => getTop());

  return { dict, headers, rows };
}
