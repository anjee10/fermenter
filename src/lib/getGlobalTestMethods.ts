import { ITestMethods } from '../Feature';

// These will be globals because of the runtime
export function getGlobalTestMethods (): ITestMethods {
  return { describe, beforeAll, afterAll, test };
}
