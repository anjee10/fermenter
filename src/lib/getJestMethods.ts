import { ITestMethods } from '../GherkinTest';

export function getJestMethods (): ITestMethods {
  return { describe, beforeAll, afterAll, test };
}
