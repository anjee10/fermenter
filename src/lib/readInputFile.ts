import { readFileSync } from 'fs';
import { dirname, isAbsolute, resolve } from 'path';

export function readInputFile ({ filePath, testFilePath }: { filePath: string, testFilePath?: string }) {
  if (isAbsolute(filePath)) {
    return readFileSync(filePath, 'utf8');
  } else {
    const testDirectory = dirname(testFilePath!);

    return readFileSync(resolve(testDirectory, filePath), 'utf8');
  }
}
