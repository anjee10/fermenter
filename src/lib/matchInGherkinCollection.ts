import * as c from 'colors';
import { IGherkinAstCollections, IGherkinCollectionItemIndex, IMatch } from '../types';
import { ExpressionMatcher } from './ExpressionMatcher';

export interface IGherkinMatchCollectionParams {
  type: IGherkinAstCollections['type'];
  match: IMatch;
  collection: IGherkinAstCollections[];
  matchProperty: 'name' | 'text';
}

export function matchInGherkinCollection<
  In extends IGherkinAstCollections = IGherkinAstCollections
> ({ type, match, collection, matchProperty }: IGherkinMatchCollectionParams): In {
  const filteredByType = <IGherkinCollectionItemIndex[]> collection
    .filter((child) => child.type === type);

  // TODO: this needs to ALSO filter on `keyword`!

  const matchingChild = filteredByType
    .find((child) => matchGherkinText(child[matchProperty], match));

  if (!matchingChild) {
    throw new Error([
      `Couldn't find a match for ${c.cyan(type)} expression:`,
      `  ${c.red(match.toString())}`,
      `Possible matches:`,
      ...filteredByType.map((child) => `  ${c.yellow(child[matchProperty])}`),
      '',
    ].join('\n'));
  }

  return matchingChild as In;
}

export function matchGherkinText (subject: string, match: IMatch): boolean {
  const matcher = ExpressionMatcher(match);

  return !!(subject.match(matcher.regexp) || []).length;
}
