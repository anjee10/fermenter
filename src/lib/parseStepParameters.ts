import { IGherkinAstStep, IMatch } from '../types';
import { ExpressionMatcher } from './ExpressionMatcher';
import { GherkinTableReader } from './GherkinTableReader';

export function parseStepParameters (step: IGherkinAstStep, match: IMatch) {
  const { argument, text } = step;

  const matcher = ExpressionMatcher(match);
  const matches = matcher.match(text) || [];

  const simpleParams = matches.map((arg) => arg.getValue());
  const argumentParam = parseStepArgument(argument);

  const params = argumentParam
    ? [...simpleParams, argumentParam]
    : simpleParams;

  return params;
}

function parseStepArgument (argument: IGherkinAstStep['argument']) {
  if (!argument) { return; }

  if (argument.type === 'DataTable') {
    return GherkinTableReader({ rows: argument.rows });
  }

  if (argument.type === 'DocString') {
    return argument.content;
  }
}
