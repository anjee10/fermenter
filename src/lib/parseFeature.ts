import * as Gherkin from 'gherkin';
import * as isValidPath from 'is-valid-path';

import { FeatureBuilder } from '../FeatureBuilder';
import { IGherkinAst } from '../types';
import { readInputFile } from './readInputFile';

export interface IGherkinParserOutput {
  ast: IGherkinAst;
  text: string;
  featureBuilder: FeatureBuilder;
}

export interface IGherkinParserConfig {
  feature: string;
  stackIndex?: number;
}

// TODO: make this pure functional, abstract the error stack relative shit out
/** Parses a feature file from a relative filePath */
export function parseFeature ({ feature, stackIndex = 2 }: IGherkinParserConfig): IGherkinParserOutput {
  const testFilePath = new Error().stack!
    .split('\n')[stackIndex]
    .match(/\(([^:]+):/ig)![0]
    .replace(/^\(|:$/g, '');

  const text = isValidPath(feature)
    ? readInputFile({ filePath: feature, testFilePath })
    : feature;

  const parser = new Gherkin.Parser();
  const ast: IGherkinAst = parser.parse(text);

  const featureBuilder = new FeatureBuilder(ast);

  return { ast, text, featureBuilder };
}
