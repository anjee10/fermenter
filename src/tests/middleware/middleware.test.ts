import { resolve } from 'path';

import { executeTest } from '../lib';

describe('Middleware hook tests', () => {
  it('middlewareHookTest', async () => {
    const output = await executeTest(resolve(__dirname, './middlewareHookTest.ts'));

    expect(output).toMatchSnapshot();
  });
});
