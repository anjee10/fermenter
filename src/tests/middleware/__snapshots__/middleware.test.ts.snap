// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware hook tests middlewareHookTest 1`] = `
"FAIL src/tests/middleware/middlewareHookTest.ts
  Feature: Middleware COR for steps, globally
    Scenario: I have a scenario
      ✓ Given: Something is set by a middleware
      ✕ Then: Check that Something is true

  ● Feature: Middleware COR for steps, globally › Scenario: I have a scenario › Then: Check that Something is true

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: undefined

    Difference:

      Comparing two different types of values. Expected boolean but received undefined.

       7 |     })
       8 |     .Then('Check that Something is true', (state) => {
    >  9 |       expect(state.something).toBe(true);
         |                               ^
      10 |     });
      11 | });
      12 | 

      at Object.Scenario.Given.Then [as fn] (src/tests/middleware/middlewareHookTest.ts:9:31)
      at src/Feature.ts:131:28
      at src/Feature.ts:7:71
      at Object.<anonymous>.__awaiter (src/Feature.ts:3:12)
      at executeStep (src/Feature.ts:75:12)
      at src/Feature.ts:212:32
      at src/Feature.ts:7:71
      at Object.<anonymous>.__awaiter (src/Feature.ts:3:12)
      at Object.testFn (src/Feature.ts:207:33)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
error Command failed with exit code 1.
"
`;
